causallearn/utils/cit.py

```diff
@@ -15,78 +15,142 @@
 chisq = "chisq"
 gsq = "gsq"
 
-class CIT(object):
-    def __init__(self, data, method='fisherz', **kwargs):
+
+def CIT(data, method='fisherz', **kwargs):
+    '''
+    Parameters
+    ----------
+    data: numpy.ndarray of shape (n_samples, n_features)
+    method: str, in ["fisherz", "mv_fisherz", "mc_fisherz", "kci", "chisq", "gsq"]
+    kwargs: placeholder for future arguments, or for KCI specific arguments now
+    '''
+    if method == fisherz:
+        return FisherZ(data, **kwargs)
+    elif method == kci:
+        return KCI(data, **kwargs)
+    elif method in [chisq, gsq]:
+        return Chisq_or_Gsq(data, method_name=method, **kwargs)
+    elif method == mv_fisherz:
+        return MV_FisherZ(data, **kwargs)
+    elif method == mc_fisherz:
+        return MC_FisherZ(data, **kwargs)
+    else:
+        raise ValueError("Unknown method: {}".format(method))
+
+class CIT_Base(object):
+    # Base class for CIT, contains basic operations for input check and caching, etc.
+    def __init__(self, data, cache_path=None, **kwargs):
         '''
         Parameters
         ----------
-        data: numpy.ndarray of shape (n_samples, n_features)
-        method: str, in ["fisherz", "mv_fisherz", "mc_fisherz", "kci", "chisq", "gsq"]
-        kwargs: placeholder for future arguments, or for KCI specific arguments now
+        data: data matrix, np.ndarray, in shape (n_samples, n_features)
+        cache_path: str, path to save cache .json file. default as None (no io to local file).
+        kwargs: for future extension.
         '''
+        assert isinstance(data, np.ndarray), "Input data must be a numpy array."
         self.data = data
-        self.data_hash = hash(str(data))
+        self.data_hash = hashlib.md5(str(data).encode('utf-8')).hexdigest()
```

Comment:
When data is huge, this will be slow? 

And when path is None, we don't need to compute data_hash?

Reply:
It will be fast (and used only once). `np.ndarray.__str__` only returns a preview:

```bash
In [41]: data = np.random.randn(10000, 100)

In [42]: str(data)
Out[42]: '[[-1.39552534  1.37974053 -1.1619043  ...  0.13616104 -0.12120668\n  -1.00001339]\n [-0.25197878 -2.00971912  0.63008704 ... -0.97997436 -1.21297862\n   1.42272323]\n [-1.22421999  0.90022162 -1.33748472 ...  1.32908047 -1.37618144\n  -0.28312766]\n ...\n [ 1.71461535  0.10882434  0.08604805 ...  1.34678215 -2.30936746\n   0.76045509]\n [ 0.55727436  0.2203048   0.41242777 ...  0.95881301  0.58538315\n   1.26002782]\n [-0.77753666  0.53018912  0.70592259 ...  0.14847539 -0.60861808\n  -0.36093896]]'

In [43]: len(str(data))
Out[43]: 491

In [44]: timeit hashlib.md5(str(data).encode('utf-8')).hexdigest()
125 µs ± 392 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```